As a user, I want more earlier and more precise feedback about why my evaluation failed when no data was found

Given an evaluation that contains no valid no time-series data on one or both of the `observed` and `predicted` orientations
When that evaluation fails
Then it should fail as early as reasonably possible and with a reasonably precise message
So that I can avoid a generic message at the end of the evaluation pipeline about why no statistics were produced

To illustrate the issue, consider system test scenario 512. Adapt the observations contained in `pizza_near_volcano_observed_in_celsius.csv` as follows:

```
value_date,variable_name,location,location_description,measurement_unit,value
0079-08-24T13:00:00Z,🍕,volcano,🌋,℃,NaN
0079-08-24T14:00:00Z,🍕,volcano,🌋,℃,NaN
0079-08-24T15:00:00Z,🍕,volcano,🌋,℃,NaN
```

Then run the existing declaration as normal. As expected, the evaluation fails due to lack of data:

```
Caused by: wres.pipeline.WresProcessingException: Statistics could not be produced for any pools. This probably occurred because none of the pools contained valid pairs. Check that the declaration contains some pools whose boundaries (e.g., earliest and latest issued times, earliest and latest valid times and earliest and latest lead durations) are sufficiently broad to capture some pairs at the desired time scale for at least one feature and threshold.
	at wres.pipeline.pooling.PoolReporter.report(PoolReporter.java:230)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:798)
	... 8 common frames omitted
```

However, this is the last possible opportunity to except, which occurs at the very end of the pipeline, once it's discovered that no pools produced statistics. For the same reason, the error cannot be very precise as it encapsulates a large number of possibilities. This could be detected much earlier, after ingest and before retrieval, and with a much more precise error message. That is the goal of this ticket.

--- a/wres-datamodel/src/wres/datamodel/time/TimeSeriesSlicer.java
+++ b/wres-datamodel/src/wres/datamodel/time/TimeSeriesSlicer.java
@@ -31,6 +31,7 @@ import java.util.stream.Collectors;
 
 import wres.config.components.DataType;
 import wres.config.components.TimeInterval;
+import wres.datamodel.MissingValues;
 import wres.datamodel.types.Climatology;
 import wres.datamodel.types.Ensemble;
 import wres.datamodel.types.Ensemble.Labels;
@@ -1488,6 +1489,46 @@ public final class TimeSeriesSlicer
             }
             return TimeSeries.of( series.getMetadata(), filtered );
         };
+    }
+
+    /**
+     * Determines whether the time-series has one or more values that do not correspond to the
+     * {@link MissingValues#DOUBLE}.
+     *
+     * @param timeSeries the time-series
+     * @return whether there are non-missing values present
+     * @throws NullPointerException if the time-series is null
+     */
+
+    public static boolean hasNonMissingValues( TimeSeries<Double> timeSeries )
+    {
+        Objects.requireNonNull( timeSeries );
+
+        return timeSeries.getEvents()
+                         .stream()
+                         .mapToDouble( Event::getValue )
+                         .anyMatch( MissingValues::isNotMissingValue );
+    }
+
+    /**
+     * Determines whether the time-series has one or more values that do not correspond to the
+     * {@link MissingValues#DOUBLE}. Equivalent to {@link TimeSeriesSlicer#hasNonMissingValues(TimeSeries)} for an
+     * ensemble time-series.
+     *
+     * @param timeSeries the time-series
+     * @return whether there are non-missing values present
+     * @throws NullPointerException if the time-series is null
+     */
+
+    public static boolean hasNonMissingEnsembleMemberValues( TimeSeries<Ensemble> timeSeries )
+    {
+        Objects.requireNonNull( timeSeries );
+
+        return timeSeries.getEvents()
+                         .stream()
+                         .flatMapToDouble( e -> Arrays.stream( e.getValue()
+                                                                .getMembers() ) )
+                         .anyMatch( MissingValues::isNotMissingValue );
     }
 
     /**

--- a/wres-datamodel/test/wres/datamodel/time/TimeSeriesSlicerTest.java
+++ b/wres-datamodel/test/wres/datamodel/time/TimeSeriesSlicerTest.java
@@ -27,6 +27,7 @@ import org.apache.commons.lang3.tuple.Pair;
 import org.junit.jupiter.api.Test;
 
 import wres.config.components.TimeInterval;
+import wres.datamodel.MissingValues;
 import wres.datamodel.types.Climatology;
 import wres.datamodel.types.Ensemble;
 import wres.datamodel.Slicer;
@@ -1661,6 +1662,56 @@ final class TimeSeriesSlicerTest
                    () -> assertTrue( TimeSeriesSlicer.canInferValidTimeFromReferenceTimeAndLeadDuration( b ) ),
                    () -> assertTrue( TimeSeriesSlicer.canInferValidTimeFromReferenceTimeAndLeadDuration( c ) ),
                    () -> assertFalse( TimeSeriesSlicer.canInferValidTimeFromReferenceTimeAndLeadDuration( d ) ) );
+    }
+
+    @Test
+    void hasNonMissingValuesPresent()
+    {
+        TimeSeriesMetadata metadata = getBoilerplateMetadata();
+
+        SortedSet<Event<Double>> seriesOneValues = new TreeSet<>();
+        seriesOneValues.add( Event.of( T1985_01_01T01_00_00Z, 1.0 ) );
+        seriesOneValues.add( Event.of( T1985_01_01T02_00_00Z, MissingValues.DOUBLE ) );
+        seriesOneValues.add( Event.of( T1985_01_01T03_00_00Z, 3.0 ) );
+
+        TimeSeries<Double> seriesOne = TimeSeries.of( metadata, seriesOneValues );
+
+        SortedSet<Event<Double>> seriesTwoValues = new TreeSet<>();
+        seriesTwoValues.add( Event.of( T1985_01_01T01_00_00Z, MissingValues.DOUBLE ) );
+        seriesTwoValues.add( Event.of( T1985_01_01T02_00_00Z, MissingValues.DOUBLE ) );
+        seriesTwoValues.add( Event.of( T1985_01_01T03_00_00Z, MissingValues.DOUBLE ) );
+
+        TimeSeries<Double> seriesTwo = TimeSeries.of( metadata, seriesTwoValues );
+
+        assertAll( () -> assertTrue( TimeSeriesSlicer.hasNonMissingValues( seriesOne ) ),
+                   () -> assertFalse( TimeSeriesSlicer.hasNonMissingValues( seriesTwo ) ) );
+    }
+
+    @Test
+    void hasNonMissingEnsembleMemberValuesPresent()
+    {
+        TimeSeriesMetadata metadata = getBoilerplateMetadata();
+
+        SortedSet<Event<Ensemble>> seriesOneValues = new TreeSet<>();
+        seriesOneValues.add( Event.of( T1985_01_01T01_00_00Z, Ensemble.of( 1.0, MissingValues.DOUBLE ) ) );
+        seriesOneValues.add( Event.of( T1985_01_01T02_00_00Z,
+                                       Ensemble.of( MissingValues.DOUBLE, MissingValues.DOUBLE ) ) );
+        seriesOneValues.add( Event.of( T1985_01_01T03_00_00Z, Ensemble.of( MissingValues.DOUBLE, 3.0 ) ) );
+
+        TimeSeries<Ensemble> seriesOne = TimeSeries.of( metadata, seriesOneValues );
+
+        SortedSet<Event<Ensemble>> seriesTwoValues = new TreeSet<>();
+        seriesTwoValues.add( Event.of( T1985_01_01T01_00_00Z,
+                                       Ensemble.of( MissingValues.DOUBLE, MissingValues.DOUBLE ) ) );
+        seriesTwoValues.add( Event.of( T1985_01_01T02_00_00Z,
+                                       Ensemble.of( MissingValues.DOUBLE, MissingValues.DOUBLE ) ) );
+        seriesTwoValues.add( Event.of( T1985_01_01T03_00_00Z,
+                                       Ensemble.of( MissingValues.DOUBLE ) ) );
+
+        TimeSeries<Ensemble> seriesTwo = TimeSeries.of( metadata, seriesTwoValues );
+
+        assertAll( () -> assertTrue( TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( seriesOne ) ),
+                   () -> assertFalse( TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( seriesTwo ) ) );
     }
 
     /**

--- a/wres-io/src/wres/io/ingesting/IngestResult.java
+++ b/wres-io/src/wres/io/ingesting/IngestResult.java
@@ -41,6 +41,12 @@ public interface IngestResult
     boolean requiresRetry();
 
     /**
+     * @return whether valid time-series data was ingested, i.e., some non-missing values.
+     */
+
+    boolean hasNonMissingData();
+
+    /**
      * How many times the timeseries is included in the left dataset.
      * @return Count of left dataset associations.
      */
@@ -71,6 +77,7 @@ public interface IngestResult
      * @param surrogateKey The surrogate key of the source data.
      * @param foundAlready true if found in the backing store, false otherwise
      * @param requiresRetry true if this requires retry, false otherwise
+     * @param hasNonMissingData true if non-missing data was found, false otherwise
      * @return a list with a single IngestResult in it
      */
 
@@ -78,13 +85,15 @@ public interface IngestResult
                                                   DataType dataType,
                                                   long surrogateKey,
                                                   boolean foundAlready,
-                                                  boolean requiresRetry )
+                                                  boolean requiresRetry,
+                                                  boolean hasNonMissingData )
     {
         IngestResult ingestResult = IngestResult.from( dataSource,
                                                        dataType,
                                                        surrogateKey,
                                                        foundAlready,
-                                                       requiresRetry );
+                                                       requiresRetry,
+                                                       hasNonMissingData );
         return List.of( ingestResult );
     }
 
@@ -95,13 +104,15 @@ public interface IngestResult
      * @param surrogateKey The surrogate key of the source data.
      * @param foundAlready true if found in the backing store, false otherwise
      * @param requiresRetry true if this requires retry, false otherwise
+     * @param hasNonMissingData true if non-missing data was found, false otherwise
      * @return the IngestResult
      */
     private static IngestResult from( DataSource dataSource,
                                       DataType dataType,
                                       long surrogateKey,
                                       boolean foundAlready,
-                                      boolean requiresRetry )
+                                      boolean requiresRetry,
+                                      boolean hasNonMissingData )
     {
         if ( requiresRetry && !foundAlready )
         {
@@ -112,14 +123,16 @@ public interface IngestResult
         {
             return new IngestResultNeedingRetry( dataSource,
                                                  dataType,
-                                                 surrogateKey );
+                                                 surrogateKey,
+                                                 hasNonMissingData );
         }
         else
         {
             return new IngestResultCompact( dataSource,
                                             dataType,
                                             surrogateKey,
-                                            foundAlready );
+                                            foundAlready,
+                                            hasNonMissingData );
         }
     }
 

--- a/wres-io/src/wres/io/ingesting/IngestResultCompact.java
+++ b/wres-io/src/wres/io/ingesting/IngestResultCompact.java
@@ -30,6 +30,7 @@ class IngestResultCompact implements IngestResult
     private final short baselineCount;
     private final short covariateCount;
     private final boolean foundAlready;
+    private final boolean hasNonMissingData;
     private final DataType dataType;
     private final Set<DatasetOrientation> orientations;
 
@@ -38,11 +39,13 @@ class IngestResultCompact implements IngestResult
      * @param dataType the optional data type
      * @param surrogateKey the surrogate key
      * @param foundAlready whether the time-series has been found already
+     * @param hasNonMissingData true if non-missing data was found, false otherwise
      */
     IngestResultCompact( DataSource dataSource,
                          DataType dataType,
                          long surrogateKey,
-                         boolean foundAlready )
+                         boolean foundAlready,
+                         boolean hasNonMissingData )
     {
         Objects.requireNonNull( dataSource, "Ingester must include datasource information." );
 
@@ -103,6 +106,7 @@ class IngestResultCompact implements IngestResult
         this.baselineCount = baselineCountInner;
         this.covariateCount = covariateCountInner;
         this.dataType = dataType;
+        this.hasNonMissingData = hasNonMissingData;
     }
 
     /**
@@ -163,6 +167,12 @@ class IngestResultCompact implements IngestResult
     }
 
     @Override
+    public boolean hasNonMissingData()
+    {
+        return this.hasNonMissingData;
+    }
+
+    @Override
     public String toString()
     {
         return new ToStringBuilder( this ).append( "surrogateKey", this.surrogateKey )
@@ -171,6 +181,7 @@ class IngestResultCompact implements IngestResult
                                           .append( "baselineCount", this.baselineCount )
                                           .append( "covariateCount", this.covariateCount )
                                           .append( "foundAlready", this.foundAlready )
+                                          .append( "hasNonMissingData", this.hasNonMissingData )
                                           .toString();
     }
 

--- a/wres-io/src/wres/io/ingesting/IngestResultNeedingRetry.java
+++ b/wres-io/src/wres/io/ingesting/IngestResultNeedingRetry.java
@@ -30,16 +30,19 @@ public class IngestResultNeedingRetry implements IngestResult
     private final short covariateCount;
     private final long surrogateKey;
     private final DataType dataType;
+    private final boolean hasNonMissingData;
 
     /**
      * Creates an instance.
      * @param dataSource the data source
      * @param dataType the optional data type
      * @param surrogateKey the surrogate key
+     * @param hasNonMissingData whether the data source has some valid values present
      */
     public IngestResultNeedingRetry( DataSource dataSource,
                                      DataType dataType,
-                                     long surrogateKey )
+                                     long surrogateKey,
+                                     boolean hasNonMissingData )
     {
         Objects.requireNonNull( dataSource, "Ingester must include datasource information." );
 
@@ -98,6 +101,7 @@ public class IngestResultNeedingRetry implements IngestResult
         this.baselineCount = baselineCountInner;
         this.covariateCount = covariateCountInner;
         this.dataType = dataType;
+        this.hasNonMissingData = hasNonMissingData;
     }
 
     @Override
@@ -128,6 +132,12 @@ public class IngestResultNeedingRetry implements IngestResult
     public boolean requiresRetry()
     {
         return true;
+    }
+
+    @Override
+    public boolean hasNonMissingData()
+    {
+        return this.hasNonMissingData;
     }
 
     @Override
@@ -164,6 +174,7 @@ public class IngestResultNeedingRetry implements IngestResult
                                           .append( "rightCount", this.rightCount )
                                           .append( "baselineCount", this.baselineCount )
                                           .append( "covariateCount", this.covariateCount )
+                                          .append( "hasNonMissingData", this.hasNonMissingData() )
                                           .toString();
     }
 

--- a/wres-io/src/wres/io/ingesting/SourceLoader.java
+++ b/wres-io/src/wres/io/ingesting/SourceLoader.java
@@ -96,13 +96,13 @@ public class SourceLoader
      * Reads and ingests time-series data and returns ingest results.
      * @param timeSeriesIngester the time-series ingester
      * @param systemSettings the system settings
-     * @param declaration the projectConfig for the evaluation
+     * @param declaration the declaration for the evaluation
      * @param griddedFeatures the gridded features cache to populate
      * @param readingExecutor the executor for performing reading tasks
      * @return ingest results
      * @throws NullPointerException if any required input is null
-     * @throws ReadException if the source could not be read
-     * @throws IngestException if the time-series could not be ingested
+     * @throws ReadException if the source data could not be read
+     * @throws IngestException if the time-series could not be ingested or no valid times-series data were discovered
      */
 
     public static List<IngestResult> load( TimeSeriesIngester timeSeriesIngester,
@@ -182,15 +182,11 @@ public class SourceLoader
                                        + "retries were exhausted." );
         }
 
-        List<IngestResult> composedResults = projectSources.stream()
-                                                           .toList();
+        // Check that some valid data was ingested for the observed/left and predicted/right sources, otherwise fail
+        // now. See GitHub #677
+        SourceLoader.validateDataAvailable( projectSources );
 
-        if ( composedResults.isEmpty() )
-        {
-            throw new IngestException( "No data were ingested." );
-        }
-
-        return composedResults;
+        return Collections.unmodifiableList( projectSources );
     }
 
     /**
@@ -869,6 +865,57 @@ public class SourceLoader
                         = Pair.of( orientedSource, new ArrayList<>() );
                 sources.put( source, sourcePair );
             }
+        }
+    }
+
+    /**
+     * Determines whether some (required) time-series data was ingested and throws an exception if not.
+     *
+     * @param projectSources the ingested data sources
+     * @throws IngestException if no data was ingested
+     */
+
+    private static void validateDataAvailable( List<IngestResult> projectSources )
+    {
+        if ( projectSources.isEmpty() )
+        {
+            throw new IngestException( "No time-series data were ingested. Please supply some valid data for "
+                                       + "evaluation and try again." );
+        }
+
+        boolean leftData = projectSources.stream()
+                                         .filter( result -> result.getDatasetOrientations()
+                                                                  .contains( DatasetOrientation.LEFT ) )
+                                         .anyMatch( IngestResult::hasNonMissingData );
+
+        boolean rightData = projectSources.stream()
+                                          .filter( result -> result.getDatasetOrientations()
+                                                                   .contains( DatasetOrientation.RIGHT ) )
+                                          .anyMatch( IngestResult::hasNonMissingData );
+
+        String phrase = "";
+
+        if ( !leftData && rightData )
+        {
+            phrase = "'observed'";
+        }
+        else if ( !rightData && leftData )
+        {
+            phrase = "'predicted'";
+        }
+        else if ( !leftData )
+        {
+            phrase = "'observed' and 'predicted'";
+        }
+
+        if ( !leftData || !rightData )
+        {
+            throw new IngestException( "Failed to identify any valid (non-missing) time-series data for the "
+                                       + phrase
+                                       + " data sources. Please declare some "
+                                       + phrase
+                                       + " data sources that contain valid "
+                                       + "time-series data and try again." );
         }
     }
 

--- a/wres-io/src/wres/io/ingesting/database/DatabaseTimeSeriesIngester.java
+++ b/wres-io/src/wres/io/ingesting/database/DatabaseTimeSeriesIngester.java
@@ -464,7 +464,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                     dataType,
                                                     0,
                                                     true,
-                                                    true );
+                                                    true,
+                                                    TimeSeriesSlicer.hasNonMissingValues( timeSeries ) );
         }
 
         DatabaseLockManager innerLockManager = this.getLockManager();
@@ -496,13 +497,17 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                            dataType,
                                                            source.getId(),
                                                            true,
-                                                           true );
+                                                           true,
+                                                           TimeSeriesSlicer.hasNonMissingValues( timeSeries ) );
             }
         }
         // Source was not inserted by this thread, so this is an existing source. But was it completed or abandoned?
         else
         {
-            results = this.finalizeExistingSource( source, dataType, dataSource );
+            results = this.finalizeExistingSource( source,
+                                                   dataType,
+                                                   dataSource,
+                                                   TimeSeriesSlicer.hasNonMissingValues( timeSeries ) );
         }
 
         return results;
@@ -531,7 +536,11 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                    source.getId() );
 
         // Finalize, which marks the source complete
-        return this.finalizeNewSource( source, dataType, latches, dataSource );
+        return this.finalizeNewSource( source,
+                                       dataType,
+                                       latches,
+                                       dataSource,
+                                       TimeSeriesSlicer.hasNonMissingValues( timeSeries ) );
     }
 
     /**
@@ -628,7 +637,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                     dataType,
                                                     0,
                                                     true,
-                                                    true );
+                                                    true,
+                                                    TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( timeSeries ) );
         }
 
         DatabaseLockManager innerLockManager = this.getLockManager();
@@ -656,17 +666,22 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                               dataSource );
 
                 // Busy, retry again later
+                boolean nonMissingValues = TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( timeSeries );
                 results = IngestResult.singleItemListFrom( dataSource,
                                                            dataType,
                                                            source.getId(),
                                                            true,
-                                                           true );
+                                                           true,
+                                                           nonMissingValues );
             }
         }
         // Source was not inserted by this thread, so this is an existing source. But was it completed or abandoned?
         else
         {
-            results = this.finalizeExistingSource( source, dataType, dataSource );
+            results = this.finalizeExistingSource( source,
+                                                   dataType,
+                                                   dataSource,
+                                                   TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( timeSeries ) );
         }
 
         return results;
@@ -694,8 +709,10 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                timeSeries,
                                                source.getId() );
 
+        boolean nonMissing = TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( timeSeries );
+
         // Finalize, which marks the source complete
-        return this.finalizeNewSource( source, dataType, latches, dataSource );
+        return this.finalizeNewSource( source, dataType, latches, dataSource, nonMissing );
     }
 
     /**
@@ -804,7 +821,9 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                         null,
                                                         sourceDetails.getId(),
                                                         false,
-                                                        !completed );
+                                                        !completed,
+                                                        // Assume some data present until proven otherwise at read time
+                                                        true );
             }
 
             // Not present, so save it
@@ -960,13 +979,15 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
      * @param dataType the data type
      * @param latches the latches used to coordinate with other ingest tasks
      * @param dataSource the data source
+     * @param nonMissingValuesPresent whether the source has some valid data
      * @return the ingest results
      */
 
     private List<IngestResult> finalizeNewSource( SourceDetails source,
                                                   DataType dataType,
                                                   Set<Pair<CountDownLatch, CountDownLatch>> latches,
-                                                  DataSource dataSource )
+                                                  DataSource dataSource,
+                                                  boolean nonMissingValuesPresent )
     {
         /* Mark the given source completed because the caller was in charge of
          * ingest and needs to mark it so.
@@ -1040,7 +1061,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                 dataType,
                                                 source.getId(),
                                                 false,
-                                                false );
+                                                false,
+                                                nonMissingValuesPresent );
     }
 
     /**
@@ -1051,12 +1073,14 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
      * @param source the source details
      * @param dataType the data type
      * @param dataSource the data source
+     * @param nonMissingValuesPresent whether the source has some valid data present
      * @return the ingest results
      */
 
     private List<IngestResult> finalizeExistingSource( SourceDetails source,
                                                        DataType dataType,
-                                                       DataSource dataSource )
+                                                       DataSource dataSource,
+                                                       boolean nonMissingValuesPresent )
     {
         // Completed, no retries needed
         if ( this.isSourceComplete( source ) )
@@ -1068,7 +1092,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                     dataType,
                                                     source.getId(),
                                                     true,
-                                                    false );
+                                                    false,
+                                                    nonMissingValuesPresent );
         }
         // Started, but not completed (unless in the interim) and not in progress, so remove and retry later
         else
@@ -1082,7 +1107,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                         dataType,
                                                         source.getId(),
                                                         true,
-                                                        false );
+                                                        false,
+                                                        nonMissingValuesPresent );
             }
 
             long surrogateKey = this.getSurrogateKey( source.getHash(), dataSource.getUri() );
@@ -1133,7 +1159,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
                                                     dataType,
                                                     source.getId(),
                                                     true,
-                                                    true );
+                                                    true,
+                                                    nonMissingValuesPresent );
         }
     }
 
@@ -1369,12 +1396,6 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
             throw new IngestException( "At least one reference datetime is required until we refactor TSV to "
                                        + "permit zero." );
         }
-
-        //        if ( rowCount != 1 )
-        //        {
-        //            throw new IngestException( "Exactly one reference datetime is required until we allow callers to "
-        //                                       + "declare which one to use at evaluation time." );
-        //        }
 
         List<String[]> rows = new ArrayList<>( rowCount );
 

--- a/wres-io/src/wres/io/ingesting/database/GriddedMetadataSaver.java
+++ b/wres-io/src/wres/io/ingesting/database/GriddedMetadataSaver.java
@@ -197,7 +197,9 @@ class GriddedMetadataSaver implements Callable<List<IngestResult>>
                                                     null,
                                                     griddedSource.getId(),
                                                     !griddedSource.performedInsert(),
-                                                    !complete );
+                                                    !complete,
+                                                    // Assume data is present as no time-series are ingested here
+                                                    true );
         }
         finally
         {

--- a/wres-io/src/wres/io/ingesting/memory/InMemoryTimeSeriesIngester.java
+++ b/wres-io/src/wres/io/ingesting/memory/InMemoryTimeSeriesIngester.java
@@ -49,6 +49,9 @@ public class InMemoryTimeSeriesIngester implements TimeSeriesIngester
         Objects.requireNonNull( timeSeriesTuple );
         Objects.requireNonNull( outerSource );
 
+        // Whether there is non-missing data present
+        boolean hasNonMissingValues = false;
+
         // Close the stream on completion
         try ( timeSeriesTuple )
         {
@@ -82,6 +85,9 @@ public class InMemoryTimeSeriesIngester implements TimeSeriesIngester
                         this.timeSeriesStoreBuilder.addSingleValuedSeries( nextTuple.getSingleValuedTimeSeries(),
                                                                            linkOrientation );
                     }
+
+                    hasNonMissingValues = this.hasNonMissingValues( hasNonMissingValues,
+                                                                    TimeSeriesSlicer.hasNonMissingValues( timeSeries ) );
                 }
 
                 // Ensemble time-series?
@@ -99,12 +105,35 @@ public class InMemoryTimeSeriesIngester implements TimeSeriesIngester
                         this.timeSeriesStoreBuilder.addEnsembleSeries( nextTuple.getEnsembleTimeSeries(),
                                                                        linkOrientation );
                     }
+
+                    boolean testHasNonMissingValues = TimeSeriesSlicer.hasNonMissingEnsembleMemberValues( timeSeries );
+                    hasNonMissingValues = this.hasNonMissingValues( hasNonMissingValues,
+                                                                    testHasNonMissingValues );
                 }
             }
 
             // Arbitrary surrogate key, since this is an in-memory ingest
-            return List.of( new IngestResultInMemory( outerSource, dataType ) );
+            return List.of( new IngestResultInMemory( outerSource, dataType, hasNonMissingValues ) );
         }
+    }
+
+    /**
+     * Indicates whether non-missing values are present based on the existing state and new state. If the existing
+     * state indicates there are non-missing values, always returns {@code true}, otherwise returns the new state.
+     *
+     * @param existingState the existing state
+     * @param newState the new state
+     * @return whether there are non-missing values present
+     */
+
+    private boolean hasNonMissingValues( boolean existingState, boolean newState )
+    {
+        if ( !existingState )
+        {
+            return newState;
+        }
+
+        return true;
     }
 
     /**

--- a/wres-io/src/wres/io/ingesting/memory/IngestResultInMemory.java
+++ b/wres-io/src/wres/io/ingesting/memory/IngestResultInMemory.java
@@ -25,12 +25,13 @@ public class IngestResultInMemory implements IngestResult
      *
      * @param dataSource the data source
      * @param dataType the optional data type
+     * @param hasNonMissingData whether the data source has valid data present
      */
-    public IngestResultInMemory( DataSource dataSource, DataType dataType )
+    public IngestResultInMemory( DataSource dataSource, DataType dataType, boolean hasNonMissingData )
     {
-        Objects.requireNonNull( dataSource, "Ingester must include datasource information." );
+        Objects.requireNonNull( dataSource, "Ingester requires datasource information." );
 
-        this.innerResult = new IngestResultNeedingRetry( dataSource, dataType, 1 );
+        this.innerResult = new IngestResultNeedingRetry( dataSource, dataType, 1, hasNonMissingData );
     }
 
     @Override
@@ -61,6 +62,12 @@ public class IngestResultInMemory implements IngestResult
     public boolean requiresRetry()
     {
         return false;
+    }
+
+    @Override
+    public boolean hasNonMissingData()
+    {
+        return this.innerResult.hasNonMissingData();
     }
 
     @Override

--- a/wres-io/test/wres/io/ingesting/InMemoryTimeSeriesIngesterTest.java
+++ b/wres-io/test/wres/io/ingesting/InMemoryTimeSeriesIngesterTest.java
@@ -1,0 +1,71 @@
+package wres.io.ingesting;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import wres.config.components.DataType;
+import wres.config.components.DatasetOrientation;
+import wres.datamodel.MissingValues;
+import wres.datamodel.time.TimeSeries;
+import wres.datamodel.time.TimeSeriesSlicer;
+import wres.datamodel.time.TimeSeriesStore;
+import wres.io.TestData;
+import wres.io.ingesting.memory.InMemoryTimeSeriesIngester;
+import wres.reading.DataSource;
+import wres.reading.TimeSeriesTuple;
+
+/**
+ * Tests the {@link InMemoryTimeSeriesIngesterTest}.
+ *
+ * @author James Brown
+ */
+class InMemoryTimeSeriesIngesterTest
+{
+    @Test
+    void testIngest()
+    {
+        TimeSeriesStore.Builder builder = new TimeSeriesStore.Builder();
+        TimeSeriesIngester ingester = InMemoryTimeSeriesIngester.of( builder );
+
+        TimeSeries<Double> timeSeries = TestData.generateTimeSeriesDoubleWithNoReferenceTimes();
+        DataSource dataSource = TestData.generateDataSource( DatasetOrientation.LEFT, DataType.SIMULATIONS );
+        TimeSeriesTuple testTuple = TimeSeriesTuple.ofSingleValued( timeSeries, dataSource );
+        List<IngestResult> results = ingester.ingest( Stream.of( testTuple ), dataSource );
+
+        TimeSeriesStore store = builder.build();
+        List<TimeSeries<Double>> actual = store.getSingleValuedSeries( DatasetOrientation.LEFT )
+                                               .toList();
+        List<TimeSeries<Double>> expected = List.of( timeSeries );
+
+        assertAll( () -> assertEquals( expected, actual ),
+                   () -> assertTrue( results.get( 0 ).hasNonMissingData() ) );
+    }
+
+    @Test
+    void testIngestWithNoData()
+    {
+        TimeSeriesStore.Builder builder = new TimeSeriesStore.Builder();
+        TimeSeriesIngester ingester = InMemoryTimeSeriesIngester.of( builder );
+
+        TimeSeries<Double> source = TestData.generateTimeSeriesDoubleWithNoReferenceTimes();
+        TimeSeries<Double> timeSeries = TimeSeriesSlicer.transform( source, e -> MissingValues.DOUBLE, m -> m );
+        DataSource dataSource = TestData.generateDataSource( DatasetOrientation.LEFT, DataType.SIMULATIONS );
+        TimeSeriesTuple testTuple = TimeSeriesTuple.ofSingleValued( timeSeries, dataSource );
+        List<IngestResult> results = ingester.ingest( Stream.of( testTuple ), dataSource );
+
+        TimeSeriesStore store = builder.build();
+        List<TimeSeries<Double>> actual = store.getSingleValuedSeries( DatasetOrientation.LEFT )
+                                               .toList();
+        List<TimeSeries<Double>> expected = List.of( timeSeries );
+
+        assertAll( () -> assertEquals( expected, actual ),
+                   () -> assertFalse( results.get( 0 ).hasNonMissingData() ) );
+    }
+}

--- a/wres-io/test/wres/io/ingesting/SourceLoaderTest.java
+++ b/wres-io/test/wres/io/ingesting/SourceLoaderTest.java
@@ -13,6 +13,7 @@ import org.junit.jupiter.api.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
@@ -33,7 +34,7 @@ class SourceLoaderTest
         CompletableFuture<String> f = CompletableFuture.supplyAsync( failingTask::call );
         CompletableFuture<Object> result = SourceLoader.doAllOrException( List.of( f ) );
         Throwable cause = assertThrows( CompletionException.class, result::join ).getCause();
-        assertTrue( cause instanceof SourceLoaderTest.DummyException );
+        assertInstanceOf( DummyException.class, cause );
     }
 
     @Test
@@ -93,7 +94,7 @@ class SourceLoaderTest
         CompletableFuture<Object> result = SourceLoader.doAllOrException( List.of( f1, f2 ) );
         Throwable cause = assertThrows( CompletionException.class, result::join ).getCause();
         Instant end = Instant.now();
-        assertTrue( cause instanceof SourceLoaderTest.DummyException );
+        assertInstanceOf( DummyException.class, cause );
         Duration executionDuration = Duration.between( start, end );
         assertTrue( executionDuration.toMillis() >= shorterDuration.toMillis()
                     && executionDuration.toMillis() < longerDuration.toMillis(),
@@ -131,7 +132,7 @@ class SourceLoaderTest
             try
             {
                 LOGGER.debug( "I am sleeping for {}", duration );
-                Thread.sleep( duration.toMillis() );
+                Thread.sleep( duration.toMillis() );  // NOSONAR
                 LOGGER.debug( "I am done sleeping for {}", duration );
             }
             catch ( InterruptedException ie )

--- a/wres-io/test/wres/io/retrieving/database/SingleValuedGriddedRetrieverTest.java
+++ b/wres-io/test/wres/io/retrieving/database/SingleValuedGriddedRetrieverTest.java
@@ -97,14 +97,14 @@ class SingleValuedGriddedRetrieverTest
     private AutoCloseable mocks;
 
     @BeforeAll
-    public static void oneTimeSetup()
+    static void oneTimeSetup()
     {
         // Set the JVM timezone for use by H2. Needs to happen before anything else
         TimeZone.setDefault( TimeZone.getTimeZone( "UTC" ) );
     }
 
     @BeforeEach
-    public void setup() throws SQLException, LiquibaseException
+    void setup() throws SQLException, LiquibaseException
     {
         this.mocks = MockitoAnnotations.openMocks( this );
         // Create the database and connection pool
@@ -279,7 +279,8 @@ class SingleValuedGriddedRetrieverTest
                                                null );
             IngestResult ingestResult = new IngestResultNeedingRetry( source,
                                                                       DataType.SINGLE_VALUED_FORECASTS,
-                                                                      sourceId );
+                                                                      sourceId,
+                                                                      true );
 
             ingestResults.add( ingestResult );
 
@@ -318,7 +319,8 @@ class SingleValuedGriddedRetrieverTest
         IngestResult leftResult = new IngestResultNeedingRetry( leftSource,
                                                                 DataType.OBSERVATIONS,
                                                                 ingestResults.get( 0 )
-                                                                             .getSurrogateKey() );
+                                                                             .getSurrogateKey(),
+                                                                true );
         ingestResults.add( leftResult );
 
         // Add a project

